### ISSUE TITLE
[Rust] Macro improperly marked as invalid

### ISSUE THREAD

**Issue by Rapptz** [sublimehq/Packages#3904]
### What happened?

The following macro gives an erroneous `source.rust meta.macro.rust invalid.illegal.rust` scope:

```rust
macro_rules! test {
    ($($name:ident),+$(,)?) => {
        pub struct Foo {
            $(
                pub $name : Option<String>
            ),+
        }

        pub async fn thing() -> () {}
    }
}
```

![image](https://github.com/sublimehq/Packages/assets/1695103/2e84d4bd-230a-44c0-a2e7-39497d913446)

From a cursory view in the repository and from testing it seems this is applied because of this rule https://github.com/sublimehq/Packages/blob/9ba6b939bb33625cd6b92fd64ac89b4b0ace2a8c/Rust/Rust.sublime-syntax#L1071-L1080

Adding the `;` after the `struct` definition *does* make the scope go away but ironically it leads to invalid code since you can't have `;` after a struct definition:

```
error: expected item, found `;`
  --> src/main.rs:15:2
   |
15 | };
   |  ^ help: remove this semicolon
   |
   = help: braced struct declarations are not followed by a semicolon
```

**Comment by FichteFoll** [comment]
The issue was a bit more involved than you initially suspected because the problem was that macro repetitions inside transcribers where basically unsupported. The linked PR #3912 fixes this particular issue for structs, but due to a lack of real world examples for other block statements I didn't see if they exhibit the same or similar issues.

If you have some more complex examples for macros, that would be much appreciated.

**Comment by Rapptz** [comment]
At first glance, I couldn't find any other macros that error out (that use repetitions) outside of this (partially censored) one:

```rs
#[macro_export]
macro_rules! metadata {
    (
        namespace $ns:literal;
        $(
            $name:ident: $ity:ty
        ),+$(,)?
    ) => {
        #[derive(Debug, Clone, Default, Eq, PartialEq)]
        pub struct Metadata {
            $(
                $name: Option<$ity>
            ),+
        }

        impl $crate::database::Metadata for Metadata {
            const NAMESPACE: &'static str = $ns;
        }
    }
}
```

The others that I use that have repetitions currently do not error, e.g.
```rs
#[macro_export]
macro_rules! header_map {
    ($($name:literal => $value:expr),* $(,)*) => {{
        let mut headers = reqwest::header::HeaderMap::with_capacity($crate::__count!($($name)*));
        $(
            headers.insert(
                reqwest::header::HeaderName::from_static($name),
                reqwest::header::HeaderValue::from_static($value),
            );
        )*

        headers
    }};
}
```

**Comment by FichteFoll** [comment]
Yeah, it's (multiple) macro transcribers inside other blocks such as a `struct` definition that are a problem here (or similarly for enums). Since they can appear just about everywhere, I'd prefer a generic solution on the syntax definition side over a specific one, but it may just not be that easy to realize because of state tracking … Will need to ponder about it a bit.
